File: day7/haskell/part2.hs
import System.Environment (getArgs)
import qualified Data.Map as Map

main :: IO ()
main = do
    args <- getArgs
    let filename = if null args then "input.txt" else head args
    content <- readFile filename
    let grid = lines content
        cols = length (head grid)
        startCol = head [c | c <- [0..cols-1], grid !! 0 !! c == 'S']
        result = simulate grid (Map.singleton (0, startCol) 1) 0
    print result

simulate :: [String] -> Map.Map (Int, Int) Int -> Int -> Int
simulate grid particles exited
    | Map.null particles = exited
    | otherwise = simulate grid newParticles (exited + exitedThisRound)
  where
    results = [moveDown grid (r, c) count | ((r, c), count) <- Map.toList particles]
    newParticlesList = concatMap fst results
    exitedThisRound = sum (map snd results)
    newParticles = Map.fromListWith (+) newParticlesList

moveDown :: [String] -> (Int, Int) -> Int -> ([((Int, Int), Int)], Int)
moveDown grid (r, c) count = go (r + 1)
  where
    rows = length grid
    cols = length (head grid)

    go row
        | row >= rows = ([], count)
        | grid !! row !! c == '^' = (leftRight, 0)
        | otherwise = go (row + 1)
      where
        leftRight = [((row, c - 1), count) | c > 0] ++ [((row, c + 1), count) | c < cols - 1]

File: day7/go/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}

	file, _ := os.Open(filename)
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	rows := len(grid)
	cols := len(grid[0])

	// Find S
	startCol := 0
	for c := 0; c < cols; c++ {
		if grid[0][c] == 'S' {
			startCol = c
			break
		}
	}

	// Track particle counts at each position
	type pos struct{ r, c int }
	particles := map[pos]int{{0, startCol}: 1}
	totalTimelines := 0

	for len(particles) > 0 {
		newParticles := map[pos]int{}

		for p, count := range particles {
			r, c := p.r, p.c
			hitSplitter := false

			// Move down
			for r = r + 1; r < rows; r++ {
				if grid[r][c] == '^' {
					hitSplitter = true
					if c > 0 {
						newParticles[pos{r, c - 1}] += count
					}
					if c < cols-1 {
						newParticles[pos{r, c + 1}] += count
					}
					break
				}
			}

			if !hitSplitter {
				totalTimelines += count
			}
		}

		particles = newParticles
	}

	fmt.Println(totalTimelines)
}
